src/background.rs

```diff
@@ -0,0 +1,476 @@
+        std::thread::Builder::new()
+            .name("background-toplevel-updates".into())
+            .spawn(move || Background::toplevel_signal(toplevel_signal, toplevel_tx))
+            .expect("Spawning toplevels update thread should succeed");
+
+        Self {
+            wayland_helper,
+            tx,
+            rx_conf,
+        }
+    }
+
+    /// Trigger [`Background::running_applications_changed`] on toplevel updates
+    fn toplevel_signal(signal: Arc<(Mutex<bool>, Condvar)>, tx: mpsc::Sender<subscription::Event>) {
+        loop {
+            let (lock, cvar) = &*signal;
+            let mut updated = lock.lock().unwrap();
+
+            log::debug!("Waiting for toplevel updates");
+            while !*updated {
+                updated = cvar.wait(updated).unwrap();
+            }
+
+            log::debug!("Emitting RunningApplicationsChanged in response to toplevel updates");
+            debug_assert!(*updated);
+            *updated = false;
+            if let Err(e) = tx.blocking_send(subscription::Event::BackgroundToplevels) {
+                log::warn!("Failed sending event to trigger RunningApplicationsChanged: {e:?}");
+            }
+        }
+    }
+
+    /// Write `desktop_entry` to path `launch_entry`.
+    ///
+    /// The primary purpose of this function is to ease error handling.
+    async fn write_autostart(
+        autostart_entry: &Path,
+        desktop_entry: &freedesktop_desktop_entry::DesktopEntry<'_>,
+    ) -> io::Result<()> {
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .mode(0o644)
+            .open(&autostart_entry)
+            .map_ok(tokio::io::BufWriter::new)
+            .await?;
+
+        file.write_all(desktop_entry.to_string().as_bytes()).await?;
+        /// Shouldn't be needed, but the file never seemed to flush to disk until I did it manually
+        file.flush().await
+    }
+}
+
+#[zbus::interface(name = "org.freedesktop.impl.portal.Background")]
+impl Background {
+    /// Status on running apps (active, running, or background)
+    async fn get_app_state(&self) -> fdo::Result<AppStates> {
+        let apps: Vec<_> = self
+            .wayland_helper
+            .toplevels()
+            .into_iter()
+            .map(|(_, info)| {
+                let status = if info
+                    .state
+                    .contains(&zcosmic_toplevel_handle_v1::State::Activated)
+                {
+                    AppStatus::Active
+                } else if !info.state.is_empty() {
+                    AppStatus::Running
+                } else {
+                    // xxx Is this the correct way to determine if a program is running in the
+                    // background? If a toplevel exists but isn't running, activated, et cetera,
+                    // then it logically must be in the background (?)
+                    AppStatus::Background
```

Comment:
I don't think this is right. A toplevel is a window, this state seems to describe applications without a window. So the logic should be: If any toplevel exists, then the app is either Running or Active. We currently don't have a protocol to enumerate applications without a window.

Reply:
Thanks for the review and help!

I checked how GNOME handles GetAppState. GNOME exposes a [D-Bus interface](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/data/dbus-interfaces/org.gnome.Shell.Introspect.xml) that allows retrieving the list of running applications. It also has a signal for app updates.

Its Background portal [uses that interface](https://gitlab.gnome.org/GNOME/xdg-desktop-portal-gnome/-/blob/main/src/background.c?ref_type=heads#L57-113) to check app statuses. An app that is running without windows is in the background. Apps with windows have their toplevel status checked (active, running).

KDE [doesn't seem to check](https://github.com/KDE/xdg-desktop-portal-kde/blob/master/src/background.cpp#L233-L259) if a program is in the background at all. :thinking: I find this confusing.

The code for the GNOME background portal is pretty straightforward, but I'll have to check if there is a way to get running apps easily via COSMIC.

Reply:
There isn't because we don't track "running applications" (which KDE probably also isn't doing), because doing so isn't exactly easy. Any running binary could be an "app".

I assume what Gnome is doing internally is:
1. Handle flatpaks as apps
3. Use app-scopes to identify apps it launched through one of it's interfaces (still doesn't help with stuff started e.g. via a terminal).

You can do similar things on cosmic, as we also set the corresponding app-scopes. (e.g. `systemctl --user list-units -t scope | grep -e "app-flatpak" -e "app-cosmic"`)

You would use the systemd dbus interface though to enumerate the scopes, filter for `app-cosmic` and `app-flatpak` and pull the `app_id` from the name (the third component). Afterwards you can lookup the desktop-file, if you need anything more.

Reply:
Thanks again for the help! I think this is ready again. I'm actively working on this patch again, so any future changes should be implemented faster. :grin: